### PR TITLE
OCPBUGS-66129: Implementation of pre-flight SFTP credential validation

### PR DESCRIPTION
**Pull request by neha037**
## Summary

Implements pre-flight SFTP credential validation to prevent wasteful resource creation from invalid credentials. Validation occurs **before** job creation, providing immediate error feedback and preventing retry loops.

## Problem

Reference Jire ticket - https://issues.redhat.com/browse/OCPBUGS-66129

When invalid SFTP credentials are provided:
- Job is created and fails repeatedly (up to `backoffLimit` retries)
- Each retry wastes cluster resources (CPU, memory, pods)
- Users receive delayed, cryptic error messages
- Unnecessary API server load from pod churn

## Solution

**Pre-flight validation** that:
- ✅ Tests SFTP connection before job creation
- ✅ Distinguishes transient errors (network issues) from permanent failures (bad credentials)
- ✅ Requeues for transient errors, fails fast for permanent issues
- ✅ Provides clear, actionable error messages in MustGather status

### New Files

**`validation.go`** (291 lines)
- `validateSFTPCredentials()`: Main validation function with 10s timeout
- `testSFTPConnection()`: SSH/SFTP connection test with context cancellation
- `TransientError`: Wrapper to distinguish retriable errors
- IPv4/IPv6 support with auto-port detection
- SSH connection leak prevention on timeout/cancellation

**`validation_test.go`** (450 lines)
- 62+ test cases covering edge cases
- Mock infrastructure for network-free testing
- Tests for: empty hosts, missing fields, IPv6, timeouts, API errors, context cancellation

### Modified Files

**`mustgather_controller.go`** (L175-207)
```go
// Validate SFTP credentials before creating job
validationErr := validateSFTPCredentials(ctx, k8sClient, secretRef, host, namespace)
if validationErr != nil {
    if IsTransientError(validationErr) {
        return reconcile.Result{Requeue: true}, validationErr  // Retry transient errors
    }
    // Permanent failure: update status and stop
    instance.Status.Status = "Failed"
    instance.Status.Reason = fmt.Sprintf("SFTP validation failed: %v", validationErr)
    return reconcile.Result{}, nil  // No job creation
}
```

## Error Handling

| Error Type                          | Behavior                                      | Example                          |
|-------------------------------------|-----------------------------------------------|----------------------------------|
| **Transient** (API timeout, network) | Requeue for retry                             | API server timeout               |
| **Permanent** (invalid credentials)  | Set status to "Failed", no job creation       | Wrong password, missing username |
| **Configuration** (empty host)       | Fail with clear error message                 | Empty SFTP host field            |

## Breaking Changes

❌ **None** - Backward compatible:

### CODE REVIEW

File: controllers/mustgather/validation.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+// is vulnerable to MITM attacks and NOT RECOMMENDED for production use.
+func validateSFTPCredentials(
+	ctx context.Context,
+	k8sClient client.Client,
+	secretRef corev1.LocalObjectReference,
+	host string,
+	namespace string,
+) error {
+	// Validate host is not empty before attempting connection
+	if strings.TrimSpace(host) == "" {
```
Comment (praveencodes) [inline controllers/mustgather/validation.go]: I think this check here is unnecessary because if the user doesn't give the Host value in the CR, it will take the default value "sftp.access.redhat.com". Refer [here](https://github.com/openshift/must-gather-operator/blob/12bf88e649689b2b686e1fb93ba4c25c1aae5009/api/v1alpha1/mustgather_types.go#L86).

File: controllers/mustgather/validation.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+		return fmt.Errorf("SFTP host cannot be empty")
+	}
+
+	// Retrieve the secret
+	secret := &corev1.Secret{}
+	err := k8sClient.Get(ctx, types.NamespacedName{
```
Comment (praveencodes) [inline controllers/mustgather/validation.go]: We are already getting the secret [here](https://github.com/openshift/must-gather-operator/blob/12bf88e649689b2b686e1fb93ba4c25c1aae5009/controllers/mustgather/mustgather_controller.go#L162). So, No need to fetch it again. We can do the username and password empty check there. If they are not empty, we can pass it as arguments to the validateSFTPCredentials function to validate them. 

Else remove the secret fetching and error response above and add it here itself.

File: controllers/mustgather/validation.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+	}
+	if !passwordExists || len(password) == 0 {
+		return fmt.Errorf("SFTP credentials secret '%s' is missing required field 'password'", secretRef.Name)
+	}
+
+	// Retrieve optional host_key for verification
+	// If not provided, testSFTPConnection will use InsecureIgnoreHostKey (NOT RECOMMENDED for production)
+	// For production use, always provide host_key to prevent MITM attacks
+	hostKeyData := ""
+	if hostKey, hostKeyExists := secret.Data["host_key"]; hostKeyExists {
```
Comment (praveencodes) [inline controllers/mustgather/validation.go]: What is this "host-key"?

File: controllers/mustgather/mustgather_controller.go [hunk 4]
```diff
@@ -171,6 +171,39 @@ func (r *MustGatherReconciler) Reconcile(ctx context.Context, request reconcile.
...
+				if IsTransientError(validationErr) {
+					reqLogger.Info("SFTP validation encountered transient error, requeuing", "error", validationErr)
+					return reconcile.Result{Requeue: true}, validationErr
+				}
+
+				// Permanent validation failure - update status and don't requeue
+				reqLogger.Error(validationErr, "SFTP credential validation failed permanently")
+				instance.Status.Status = "Failed"
+				instance.Status.Completed = true
+				instance.Status.Reason = fmt.Sprintf("SFTP validation failed: %v", validationErr)
```
[the remark this answers is not in the document]
Comment (neha037) [reply controllers/mustgather/mustgather_controller.go]: added

File: controllers/mustgather/mustgather_controller.go [hunk 5]
```diff
@@ -171,6 +172,51 @@ func (r *MustGatherReconciler) Reconcile(ctx context.Context, request reconcile.
...
+
+			if !passwordExists || len(password) == 0 {
+				validationErr := fmt.Errorf("SFTP credentials secret '%s' is missing required field 'password'", secretName)
+				reqLogger.Error(validationErr, "SFTP credential validation failed")
+				return r.setValidationFailureStatus(ctx, instance, validationErr)
+			}
+
+			// Extract optional host_key for verification
+			hostKeyData := ""
+			if hostKey, hostKeyExists := userSecret.Data["host_key"]; hostKeyExists {
```
[the remark this answers is not in the document]
Comment (neha037) [reply controllers/mustgather/mustgather_controller.go]: host-key are not compulsory for development and testing, but in the production environment, secrets have host-key to prevent man-in-the-middle attack.  Reference - https://www.ssh.com/academy/attack/man-in-the-middle#:~:text=governments%20against%20HTTPS.-,Various%20ways%20to%20prevent%20the%20attack,helpful%20in%20achieving%20this%20goal.

My first implementation was without host key and that let to https://github.com/openshift/must-gather-operator/pull/317#discussion_r2683306735  issue

So, considering all the scenarios, I went ahead with the implementation that if someone has configuring there host  with host-key, the host-key will be used, if not ( say for testing purposes) ssh.InsecureIgnoreHostKey() will be set.

File: controllers/mustgather/validation.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+		errChan <- testSFTPConnection(string(username), string(password), host)
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-validationCtx.Done():
+		return fmt.Errorf("SFTP credential validation timed out after %v", sftpValidationTimeout)
+	}
+}
```
[the remark this answers is not in the document]
Comment (neha037) [reply controllers/mustgather/validation.go]: Resolved here - https://github.com/openshift/must-gather-operator/pull/317/commits/8d098fec65007ac3146188658ffc6301e3f1e163

**Review by KeenonLee — COMMENTED** [review]
/label qe-approved
